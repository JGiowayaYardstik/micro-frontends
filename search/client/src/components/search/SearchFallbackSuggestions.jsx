import React from 'react';
import PropTypes from 'prop-types';
import qs from 'qs';
import { withI18n } from 'react-i18next';

const removeDuplicatesFromArray = suggestions =>
  suggestions.reduce((prev, curr) => {
    if (prev.indexOf(curr) < 0) {
      prev.push(curr);
    }
    return prev;
  }, []);

const getSuggestionLinks = suggestions =>
  removeDuplicatesFromArray(suggestions)
    .map(suggestion => {
      const queryParams = qs.parse(window.location.search);
      queryParams.q = suggestion;
      return (
        <a
          title={suggestion}
          key={suggestion}
          data-dmid="suggestion-link"
          href={decodeURIComponent(qs.stringify(queryParams))}>
          {suggestion}
        </a>
      );
    })
    .reduce((prev, curr) => [prev, ', ', curr]);

const SearchFallbackSuggestions = props => {
  const { suggestions, styles, t } = props;

  return (
    !(suggestions.length <= 0) && (
      <div data-dmid="search-fallback-suggestions" className={styles.suggestions}>
        {t('suggestions.label')} {getSuggestionLinks(suggestions)}?
      </div>
    )
  );
};

SearchFallbackSuggestions.propTypes = {
  suggestions: PropTypes.array,
  t: PropTypes.func,
};

export default withI18n()(SearchFallbackSuggestions);
