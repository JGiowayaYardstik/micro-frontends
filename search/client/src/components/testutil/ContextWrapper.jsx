import React from 'react';
import PropTypes from 'prop-types';
import { StyleProvider } from '@dm/style-provider';
import { DeviceProvider } from '@dm/device-provider';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import { theme as dm } from '@dm/ui-theme';

export default function ContextWrapper({ children, locale }) {
	return (
		<StyleProvider>
			<DeviceProvider>
				<I18nextProvider i18n={global.i18next} initialLanguage={locale}>
					{children}
				</I18nextProvider>
			</DeviceProvider>
		</StyleProvider>
	);
}

ContextWrapper.propTypes = {
	children: PropTypes.node,
	locale: PropTypes.string
};

export function ContextWrapperWithoutResponsive({ children, locale }) {
	return (
		<StyleProvider>
			<I18nextProvider i18n={global.i18next} initialLanguage={locale}>
				{children}
			</I18nextProvider>
		</StyleProvider>
	);
}

ContextWrapperWithoutResponsive.propTypes = {
	children: PropTypes.node,
	locale: PropTypes.string
};

export function ContextWrapperWithStore({ children, locale, theme = dm, store = global.store }) {
	return (
		<Provider store={store}>
			<StyleProvider theme={theme}>
				<I18nextProvider i18n={global.i18next} initialLanguage={locale}>
					{children}
				</I18nextProvider>
			</StyleProvider>
		</Provider>
	);
}

ContextWrapperWithStore.propTypes = {
	children: PropTypes.node,
	locale: PropTypes.string,
	store: PropTypes.object,
	theme: PropTypes.object
};
