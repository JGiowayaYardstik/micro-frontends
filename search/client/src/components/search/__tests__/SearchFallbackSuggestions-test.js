import React from 'react';
import { mount } from 'enzyme';
import { createSnapshot } from '@dm/jest-utils';
import ContextWrapper from '../../testutil/ContextWrapper';

import SearchFallbackSuggestions from '../SearchFallbackSuggestions';

describe('SearchFallbackSuggestions', () => {
  it('renders and matches snapshot (DE)', () => {
    const snapshot = createSnapshot(
      <ContextWrapper locale="de">
        <SearchFallbackSuggestions suggestions={['hello', 'world']} />
      </ContextWrapper>
    );
    expect(snapshot).toMatchSnapshot();
  });

  it('renders and matches snapshot (HU)', () => {
    const snapshot = createSnapshot(
      <ContextWrapper locale="hu">
        <SearchFallbackSuggestions suggestions={['hello', 'world']} />
      </ContextWrapper>
    );
    expect(snapshot).toMatchSnapshot();
  });

  it('renders and matches snapshot when duplicate suggestions are found', () => {
    const snapshot = createSnapshot(
      <ContextWrapper locale="de">
        <SearchFallbackSuggestions suggestions={['baldrian', 'balea', 'balea']} />
      </ContextWrapper>
    );
    expect(snapshot).toMatchSnapshot();
  });

  it('mounts and displays no suggestions when no suggestions are found', () => {
    const component = mount(
      <ContextWrapper>
        <SearchFallbackSuggestions suggestions={[]} />
      </ContextWrapper>
    );

    expect(component.find('[data-dmid="search-fallback-suggestions"]')).toHaveLength(0);
  });
});
