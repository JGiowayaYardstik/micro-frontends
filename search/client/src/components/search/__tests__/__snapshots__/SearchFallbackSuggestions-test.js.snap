// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchFallbackSuggestions renders and matches snapshot (DE) 1`] = `
Object {
  "_locale": "de",
  "_width": 1200,
  "html": "
<div data-dmid=search-fallback-suggestions className=suggestions__e925jf>
  Meinten Sie:{ }
  <a title=hello data-dmid=suggestion-link href=q=hello>
    hello
  </a>,{ }
  <a title=world data-dmid=suggestion-link href=q=world>
    world
  </a>?
</div>;
",
  "styles": "
.suggestions__e925jf {
  font-size: 1.125rem;
  font-family: dmbrand-medium, Arial, Helvetica, sans-serif;
  margin-bottom: 2.5rem;
  text-align: center;
}
",
}
`;

exports[`SearchFallbackSuggestions renders and matches snapshot (HU) 1`] = `
Object {
  "_locale": "de",
  "_width": 1200,
  "html": "
<div data-dmid=search-fallback-suggestions className=suggestions__e925jf>
  Erre gondolt:{ }
  <a title=hello data-dmid=suggestion-link href=q=hello>
    hello
  </a>,{ }
  <a title=world data-dmid=suggestion-link href=q=world>
    world
  </a>?
</div>;
",
  "styles": "
.suggestions__e925jf {
  font-size: 1.125rem;
  font-family: dmbrand-medium, Arial, Helvetica, sans-serif;
  margin-bottom: 2.5rem;
  text-align: center;
}
",
}
`;

exports[`SearchFallbackSuggestions renders and matches snapshot when duplicate suggestions are found 1`] = `
Object {
  "_locale": "de",
  "_width": 1200,
  "html": "
<div data-dmid=search-fallback-suggestions className=suggestions__e925jf>
  Meinten Sie:{ }
  <a title=baldrian data-dmid=suggestion-link href=q=baldrian>
    baldrian
  </a>,{ }
  <a title=balea data-dmid=suggestion-link href=q=balea>
    balea
  </a>?
</div>;
",
  "styles": "
.suggestions__e925jf {
  font-size: 1.125rem;
  font-family: dmbrand-medium, Arial, Helvetica, sans-serif;
  margin-bottom: 2.5rem;
  text-align: center;
}
",
}
`;
